Fix possible JMX metrics leak

devozerov opened the pull request:
Related to https://github.com/hazelcast/hazelcast/issues/16206
Added logging to understand where exactly the leak happens.

devozerov commented:
run-lab-run

devozerov commented:
run-lab-run

devozerov commented:
run-lab-run

vojtechtoman commented on hazelcast/src/test/java/com/hazelcast/internal/metrics/jmx/JmxPublisherTestHelper.java:
```diff
@@ -47,7 +48,7 @@ public JmxPublisherTestHelper(String domainPrefix) throws Exception {
     public void assertNoMBeans() {
         Set<ObjectInstance> instances = queryOurInstances();
-        assertEquals(0, instances.size());
+        assertEquals(Arrays.toString(instances.toArray()), 0, instances.size());
```
minor: `assertThat(instances, is(empty()))`?

devozerov replied:
I added `Arrays.toString(instances.toArray())` to print the list of dangling MBeans in the error message. AFAIK `assertThat` without an explicit message will not print anything, will it?

vojtechtoman replied:
It should display the contents of the collection if not empty

devozerov replied:
Implemented. Thanks for the suggestion.

devozerov commented:
@vojtechtoman @blazember Guys, please note that this is a draft, please do not approve it yet. I am trying to figure out what causes test failures.
I should have created this PR as "Draft".

devozerov commented:
@vojtechtoman @blazember Colleagues, could you please review the PR? It is in the final state now. There are no changes to the core part. Here is what was changed:
1) Since the original failure was due to JMX beans which were not unregistered by some of the earlier tests, I added a facility to check for leaked JMX beans on test instance factory shutdown/terminate
2) OOME tests were changed - now every test creates an instance with a unique name, thus minimizing the risk of leaked JMX beans (since registration and deregistration is racy in its nature)
3) Last, I fixed several places in `TestHazelcastInstanceFactory` when recently introduced `MetricRule` was not used for instances created by the factory.

vojtechtoman commented on hazelcast/src/test/java/com/hazelcast/test/TestHazelcastInstanceFactory.java:
```diff
@@ -296,33 +310,45 @@ public HazelcastInstance getInstance(Address address) {
+                registry.removeInstance(address);
+            }
+        } finally {
+            TestJmxLeakHelper.checkJmxBeans(instance.getName());
```
We have a large number of tests that don't use `TestHazelcastInstanceFactory` at all and call `Hazelcast.newHazelcastInstance()` directly (in combination with `Hazelcast.shutdownAll()` / `HazelcastInstanceFactory.shutdownAll()` / `HazelcastInstanceFactory.terminateAll()`). Shouldn't those be checked as well, somehow?

blazember replied:
I think it is possible to do this somewhere here: https://github.com/hazelcast/hazelcast/blob/86a759ac20eab87d7a249fa6c267a5f5ab4ad215/hazelcast/src/test/java/com/hazelcast/test/AbstractHazelcastClassRunner.java#L304-L310

I would prefer that for the reasons Vojtech mentioned. For example, the client tests rarely use `TestHazelcastFactory`, but clients also create MBeans under the same `com.hazelcast.*` namespace which this solution ignores.

devozerov replied:
@vojtechtoman @blazember Thank you very much for suggestions. I implemented the proposed strategy and it helped to identify the leaking class immediately - it was `MetricsServiceTest`. Changing it to serial mode along with explicit JMX cleanup seems to solve the original problem. Let's wait for Jenkins results.

blazember commented on hazelcast/src/test/java/com/hazelcast/instance/impl/AbstractOutOfMemoryHandlerTest.java:
```diff
@@ -35,8 +35,10 @@
+import java.util.concurrent.atomic.AtomicInteger;
 
 public abstract class AbstractOutOfMemoryHandlerTest extends HazelcastTestSupport {
+    private static final AtomicInteger INSTANCE_ID_COUTNER = new AtomicInteger();
```
Minor: typo in couTner.

devozerov replied:
Fixed.

blazember commented on hazelcast/src/test/java/com/hazelcast/test/TestHazelcastInstanceFactory.java:
```diff
@@ -140,7 +144,11 @@ public HazelcastInstance newHazelcastInstance(Address address, Config config) {
-            return HazelcastInstanceFactory.newHazelcastInstance(config, instanceName, nodeContext);
+            HazelcastInstance hazelcastInstance =
+                HazelcastInstanceFactory.newHazelcastInstance(config, instanceName, nodeContext);
+            registerInstance(hazelcastInstance);
```
Oops, thanks :smile:

blazember commented on hazelcast/src/test/java/com/hazelcast/test/TestJmxLeakHelper.java:
```diff
@@ -0,0 +1,98 @@
+
+    private static List<String> getActiveJmxBeans(String instanceName) {
+        // Escape instance name the same way it is escaped for member JMX service.
+        String instanceNameEscaped = JmxPublisher.escapeObjectNameValue(instanceName);
```
If we check the leaking MBeans in the runner's `AfterClass`, I think we don't need the instance name and no need for making `escapeObjectNameValue()` public.

devozerov replied:
I'll resolve this to continue a single discussion under Vojtech's comment.

devozerov commented:
@vojtechtoman @blazember @sancar Guys, could you please perform the review again? Now the branch is in the final state. Walkthrough:
1) We added `JmxLeakHelper` class which is used after `@AfterClass` for all tests and check for not unregistered JMX beans
2) The original test failure was caused by `MetricsServiceTest`, it is fixed now with explicit JMX bean deregistration after tests
3) `JmxLeakHelper` helped find several tests where we forgot to shutdown clients. Fixed
4) Finally, I revealed that `shutdown` implement of a client is racy. The shutdown might be initiated by either user or internal timeout service. The original implementation of that method used CAS to ensure that only one thread performs the actual shutdown. The problem with this approach is that another thread doing the shutdown will leave this method before all the components are closed. As a result, some dangling components of one test, specifically - JMX beans, may interfere with the next tests. To fix that problem I added a mutex to the `shutdown` method so that looser thread still waits for complete shutdown to finish before returning the control.

vojtechtoman commented on hazelcast/src/main/java/com/hazelcast/client/impl/clientside/LifecycleServiceImpl.java:
```diff
@@ -60,6 +60,7 @@
     private final BuildInfo buildInfo;
     private final ExecutorService executor;
 
+    /** Monitor which ensures that all client componets are down when shutdown() is finished. */
```
minor typo: componets.

devozerov replied:
Fixed.

vojtechtoman commented on hazelcast/src/test/java/com/hazelcast/client/cache/instance/ClientCacheThroughHazelcastInstanceTest.java:
```diff
@@ -64,7 +64,8 @@ protected HazelcastInstance createInstance(Config config) {
 
         ClientConfig clientConfig = new ClientConfig();
-        clientConfig.getConnectionStrategyConfig().getConnectionRetryConfig().setClusterConnectTimeoutMillis(1000);
+        clientConfig.getConnectionStrategyConfig().getConnectionRetryConfig().setClusterConnectTimeoutMillis(1_000);
```
Just curious: what was the motivation for this change? Can't this lead to more frequent test failures in case of hiccups?

devozerov replied:
This is what helped to find the race in the client shutdown. Also, it decreased test speed significantly. But I agree that it may lead to flaky failures in case of hiccups. So I increased the value to 10s. This is smaller than before (30s), but should be enough to avoid flakiness.

blazember commented on hazelcast/src/test/java/com/hazelcast/internal/metrics/impl/MetricsServiceTest.java:
```diff
@@ -123,6 +124,14 @@ public void setUp() {
     @After
     public void tearDown() {
+        // Destroy JMX beans created during testing.
+        MetricsService metricsService = new MetricsService(nodeEngineMock, () -> metricsRegistry);
```
I think alternatively we can move out the construction of the service from the test cases into a factory method that sets a shutdown function object field that we can call in the tearDown. This may prevent some other unwanted potential side-effects. With this approach we construct two services, but shutdown only one. For example, the service submits a scheduled task. I think it is fine for now, but it would be safer with the function proposed.

devozerov replied:
Implemented, thanks for the suggestion.

devozerov commented:
run-lab-run

devozerov commented:
run-lab-run

devozerov commented:
Repeated failure in `com.hazelcast.internal.ascii.RestClusterBadRequestTest.testBadRequest`. According to the Git log, there were some attempts to fix it recently. Works fine locally on the latest master, while it was failing on Friday. Re-run again.

devozerov commented:
There were two test failures in the last Jenkins run which are flaky, reproducible in master, and are not related to the given PR. Merging.
